### PR TITLE
Use internal reader when building SecurityIndexReaderWrapper filter query

### PR DESCRIPTION
**Pull request by jfreden**
Relates: https://github.com/elastic/elasticsearch/issues/112130

When using wildcards to match fields in a `multi_match` query within a DLS query in a role, the fields are not properly expanded because there is no index reader that can read the index mapping to expand the wildcards.

This happens because a [searcher](https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/index/query/SearchExecutionContext.java#L685) is needed to read the mappings to figure out what fields are available for an index.

The searcher is expected to be supplied by the Security plugin [when it calls](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java#L1492) `IndexModule::setReaderWrapper` to override the `IndexReader` factory. Currently it passes `null` for the searcher parameter, which makes expansions of wildcard fields impossible.

This PR changes that by providing the reader that wrapped by `SecurityIndexReaderWrapper` when the filter query is built to be able to resolve wildcard fields.

### CODE REVIEW

**Comment by tvernum** [comment]
I think
1. this should be backported to 8.15.x as well, since it's a bug
2. it would be good to get an additional reviewer from search engine, just to make sure we're not doing something dangerous.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java [hunk 1]
```diff
@@ -1484,16 +1485,10 @@ public void onIndexModule(IndexModule module) {
                 assert dlsBitsetCache.get() != null;
                 module.setReaderWrapper(
                     indexService -> new SecurityIndexReaderWrapper(
-                        shardId -> indexService.newSearchExecutionContext(
-                            shardId.id(),
-                            0,
-                            // we pass a null index reader, which is legal and will disable rewrite optimizations
-                            // based on index statistics, which is probably safer...
```
Comment (tvernum) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java:1491]: ^^ This concerns me.
We intentionally passed in a null reader in order to disable rewrite optimizations. I don't think we can blindly undo that.

Comment (javanna) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java:1491]: it is more to do with usages of `SearchExecutionContext#getIndexReader` than with `fieldExistsInIndex`. If a searcher is set we may rewrite queries to match none depending on index statistics, for instance for range queries against shards that don't have data within the specified range. That means that certain shards can be skipped entirely. They would not return results if they were to execute the query though, which means that it does not affect the returned results. Unless we rewrite against an empty reader, which is going to yield different results than executing the query against the real reader. I believe here we have the actual reader against the shard available so we should be ok?

Comment (jfreden) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java:1491]: Thanks for looking into it @javanna! 

>Unless we rewrite against an empty reader, which is going to yield different results than executing the query against the real reader. I believe here we have the actual reader against the shard available so we should be ok?

The reader wrapper is wrapping the actual reader that's used by the `indexService` from what I can see [here](https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/index/IndexService.java#L268), so it should always be the actual reader against the shard. 

As long as the result of the query doesn't change I think it should be safe from a DLS standpoint. Do you have any additional concerns @tvernum ?

**Pull request closed without merging**